include/openthread/srp_client.h

```diff
@@ -0,0 +1,501 @@
+ *    (7)  YXRRSET   Some RRset that ought not to exits, does exist  -> OT_ERROR_DUPLICATED
+ *    (8)  NXRRSET   Some RRset that ought to exist, does not exist  -> OT_ERROR_NOT_FOUND
+ *    (9)  NOTAUTH   Service is not authoritative for zone           -> OT_ERROR_SECURITY
+ *    (10) NOTZONE   A name is not in the zone                       -> OT_ERROR_PARSE
+ *    (20) BADNAME   Bad name                                        -> OT_ERROR_PARSE
+ *    (21) BADALG    Bad algorithm                                   -> OT_ERROR_SECURITY
+ *    (22) BADTRUN   Bad truncation                                  -> OT_ERROR_PARSE
+ *    Other response codes                                           -> OT_ERROR_FAILED
+ *
+ * The following errors are also possible:
+ *
+ *    OT_ERROR_RESPONSE_TIMEOUT : Timed out waiting for response from server (client would continue to retry).
+ *    OT_ERROR_INVALID_ARGS     : The provided service structure is invalid (e.g., bad service name or `otSrpTxtEntry`).
+ *    OT_ERROR_NO_BUFS          : Insufficient buffer to prepare or send the update message.
+ *
+ * Note that in case of any failure, the client continues the operation, i.e. it prepares and (re)transmits the SRP
+ * update message to the server, after some wait interval. The retry wait interval starts from the minimum value and
+ * is increased by the growth factor every failure up to the max value (please see configuration parameter
+ * `OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL` and the related ones for more details).
+ *
+ * @param[in] aError            The error (see above).
+ * @param[in] aHostInfo         A pointer to host info.
+ * @param[in] aService          The head of linked-list containing all services (excluding the ones removed). NULL if
+ *                              the list is empty.
+ * @param[in] aRemovedServices  The head of linked-list containing all removed services. NULL if the list is empty.
+ * @param[in] aContext          A pointer to an arbitrary context (provided when callback was registered).
+ *
+ */
+typedef void (*otSrpClientCallback)(otError                    aError,
+                                    const otSrpClientHostInfo *aHostInfo,
+                                    const otSrpClientService * aServices,
+                                    const otSrpClientService * aRemovedServices,
+                                    void *                     aContext);
+
+/**
+ * This function starts the SRP client operation.
+ *
+ * SRP client will prepare and send "SRP Update" message to the SRP server once all the following conditions are met:
+ *
+ *  - The SRP client is started - `otSrpClientStart()` is called.
+ *  - Host name is set - `otSrpClientSetHostName()` is called.
+ *  - At least one host IPv6 address is set - `otSrpClientSetHostName()` is called.
+ *  - At least one service is added - `otSrpClientAddService()` is called.
```

Comment:
I have been considering and going back and forth on whether to keep/add this check or not. I know it is possible/valid to register just host info.

At the end I added this check since I think this can be very useful in practical use-cases (particularly on SED type devices). User may add the host info early-on but then services may be added later (e.g. during device pairing or start of operation, ect). The idea is to not send SRP update until we have at least one service. This helps reduce the number of tx on SED type devices.

I suggest if we see in future that there is a need for registering host info alone (without any service) we add a build config option to enable this behavior (it's just removing one line of code in `CheckState()`).

Reply:
okay to keep it is.